app/code/ZP/LoyaltyProgram/Controller/Adminhtml/Program/MassDelete.php

```diff
@@ -3,52 +3,135 @@
+use ZP\LoyaltyProgram\Api\LoyaltyProgramRepositoryInterface;
+use ZP\LoyaltyProgram\Api\Model\Validators\Controller\Adminhtml\Program\MassAction\ValidatorInterface;
 
-class MassDelete extends Action implements HttpPostActionInterface
+class MassDelete extends Controller
 {
-    /**
-     * Authorization level of a basic admin session
-     *
-     * @see _isAllowed()
-     */
-    const ADMIN_RESOURCE = 'ZP_LoyaltyProgram::manage';
+    private const DELETE = 'delete';
+
+    private int $deletedProgramsCount = 0;
 
     public function __construct(
         Context $context,
-        private Filter $filter,
-        private CollectionFactory $collectionFactory,
-        private LoyaltyProgramResource $programResource
+        LoggerInterface $logger,
+        LoyaltyProgramRepositoryInterface $programRepository,
+        CustomerProgramManagement $customerProgramManagement,
+        Helper $helper,
+        ValidatorInterface $dataValidator,
+        RequestHelper $requestHelper,
+        Filter $filter,
+        CollectionFactory $collectionFactory,
+        DataPreparer $prepareData,
+        SalesRuleProgramsManagement $salesRuleProgramsManagement
     ) {
-        parent::__construct($context);
+        parent::__construct(
+            $context,
+            $logger,
+            $programRepository,
+            $customerProgramManagement,
+            $helper,
+            $dataValidator,
+            $requestHelper,
+            $filter,
+            $collectionFactory,
+            $prepareData
+        );
+
+        $this->salesRuleProgramsManagement = $salesRuleProgramsManagement;
     }
 
     public function execute()
     {
-        $collection = $this->filter->getCollection($this->collectionFactory->create());
-        $collectionSize = $collection->getSize();
+        try {
+            /** @var Redirect $resultRedirect */
+            $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            $this->checkProgramIds(self::DELETE);
 
-        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
-        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            /** @var Collection $collection */
+            $collection = $this->filter->getCollection($this->collectionFactory->create());
 
-        try {
-            /** @var LoyaltyProgram $program */
-            foreach ($collection as $program) {
-                $this->programResource->delete($program);
-            }
+            if (!$collection->getSize()) {
+                $this->messageManager->addNoticeMessage(
+                    __('Selected program(s) has(have) been probably deleted already!')
+                );
+            } else {
+                $this->validateCollectionPrograms($collection);
+                if ($this->programsForAction) {
+                    $this->programsData = $this->programsForAction;
+                    $this->beforeDelete();
```

Comment:
No we cant, because my before and after actions which are in controllers have action logic that is not applicable to our resource model. But I will try to remember this moment and we will discuss it on our next live meeting, and may in our discussion we will find the way how to do it even more functional.